### PR TITLE
Add search bar for nanonote

### PR DESCRIPTION
**Pull request by pavol47**
The search bar supports find words, higlighted words, find words with case sensitivity and only whole words.

### CODE REVIEW

**Comment by agateau** [comment]
Hi, sorry for the late reply, going to look at your code tonight.

**Comment by pavol47** [comment]
I change code according to the requirements of changes request. I hope everything is fixed. Thank you for your code review.

File: src/SearchWidget.cpp [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+}
+
+void SearchWidget::searchWord(bool selectNext, QString searchValue)
+{
+    if (searchValue.isEmpty()) {
+        searchValue = mUi->guiSearchLine->text();
+    }
+    mTextDocument = mTextEdit->toPlainText();
+    mCurrentSelectedWord = -1;
+    mPositionWords.clear();
```
Comment (agateau) [inline src/SearchWidget.cpp]: Clearing `mPositionWords` should be done by `searchPositionsWordsInDocument`, it's its job to update `mPositionWords`.

File: src/SearchWidget.cpp [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    QTextDocument *document = mTextEdit->document();
+    QTextCursor cursor(document);
+    cursor.beginEditBlock();
+    int wordStartPosition = mPositionWords.at(mCurrentSelectedWord);
+    cursor.setPosition(wordStartPosition, QTextCursor::MoveAnchor);
+    cursor.setPosition(wordStartPosition + mUi->guiSearchLine->text().size(), QTextCursor::KeepAnchor);
+    mTextEdit->setTextCursor(cursor);
+    cursor.endEditBlock();
+    setCountAndCurrentPosition();
+}
```
Comment (agateau) [inline src/SearchWidget.cpp]: There is a lot of duplication between `onNextButtonClicked()` and `onPreviousButtonClicked()`.
You should be able to factorize all the code after the update of mCurrentSelectedWord between the two methods.

File: CMakeLists.txt [hunk 3]
```diff
@@ -12,6 +12,9 @@ add_subdirectory(3rdparty/catch2)
 
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Woverloaded-virtual")
 set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_DEBUG_POSTFIX "d")
+
+set(CMAKE_BUILD_TYPE "Debug")
```
Comment (agateau) [inline CMakeLists.txt]: This looks wrong. It prevents the user running cmake from defining the build type.

File: CMakeLists.txt [hunk 4]
```diff
@@ -12,6 +12,9 @@ add_subdirectory(3rdparty/catch2)
 
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Woverloaded-virtual")
 set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_DEBUG_POSTFIX "d")
```
Comment (agateau) [inline CMakeLists.txt]: This looks unnecessary. At least it does not feel related to the search bar ;)

File: src/MainWindow.cpp [hunk 5]
```diff
@@ -136,29 +160,29 @@ void MainWindow::loadNotes()
...
+                                    "- This works for\n"
+                                    "    - nested lists\n"
+                                    "    * and asterisks\n"
+                                    "\n"
+                                    "You can also open urls using Control + click or Control + Enter while your cursor is inside a URL. You can try clicking on this one for example: https://github.com/agateau/nanonote.\n"
+                                    "\n"
+                                    "Finally, you can indent selected lines with Tab or Ctrl+I and unindent them with Shift+Tab or Ctrl+U.\n"
+                                    "\n"
+                                    "That's all there is to say, now you can erase this text and start taking notes!\n"
+                                    ));
```
Comment (agateau) [inline src/MainWindow.cpp]: format: no need to reindent this.

File: src/SearchWidget.cpp [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+{
+    emit closeSearchDialog();
+}
+
+void SearchWidget::searchLineChanged(const QString & value)
+{
+    searchWord(true, value);
+}
+
+void SearchWidget::searchPositionsWordsInDocument(const QString & searchString, bool selectNext, QTextDocument * document)
```
[the remark this answers is not in the document]
Comment (pavol47) [reply src/SearchWidget.cpp]: resolve in commit 316749c

File: src/SearchWidget.cpp [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+    cursor.setPosition(wordStartPosition + mUi->guiSearchLine->text().size(), QTextCursor::KeepAnchor);
+    mTextEdit->setTextCursor(cursor);
+    cursor.endEditBlock();
+    setCountAndCurrentPosition();
+}
+
+void SearchWidget::highLightedWords(bool highLighted)
+{
+    QList<QTextEdit::ExtraSelection> extraSelections;
+    QColor redColor = Qt::red;
```
[the remark this answers is not in the document]
Comment (pavol47) [reply src/SearchWidget.cpp]: resolve in commit 44fce30

File: src/SearchWidget.cpp [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+
+#include "TextEdit.h"
+
+SearchWidget::SearchWidget(TextEdit* textEdit, QWidget *parent )
+    : QWidget( parent )
+    , mUi(new Ui::SearchForm)
+    , mTextEdit(textEdit)
+{
+    mUi->setupUi(this);
+    mUi->guiPreviousButton->setArrowType(Qt::UpArrow);
```
[the remark this answers is not in the document]
Comment (pavol47) [reply src/SearchWidget.cpp]: resolve in commit eed2dbc

File: src/SearchWidget.cpp [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+    mUi->guiPreviousButton->setToolButtonStyle(Qt::ToolButtonTextBesideIcon);
+    mUi->guiNextButton->setArrowType(Qt::DownArrow);
+    mUi->guiNextButton->setToolButtonStyle(Qt::ToolButtonTextBesideIcon);
+
+    QObject::connect(mUi->guiNextButton, &QToolButton::clicked, this, &SearchWidget::onNextButtonClicked);
+    QObject::connect(mUi->guiPreviousButton, &QToolButton::clicked, this, &SearchWidget::onPreviousButtonClicked);
+    QObject::connect(mTextEdit, &TextEdit::textChanged, this, &SearchWidget::documentChange);
+    QObject::connect(mUi->guiSearchLine, &QLineEdit::textChanged, this, &SearchWidget::searchLineChanged);
+    QObject::connect(mUi->guiCloseButton, &QToolButton::clicked, this, &SearchWidget::closeSearch);
```
[the remark this answers is not in the document]
Comment (pavol47) [reply src/SearchWidget.cpp]: resolve in commit 8079859

File: CMakeLists.txt (hunk 3 above)
Comment (pavol47) [reply CMakeLists.txt]: resolve in commit 9344a35

File: CMakeLists.txt (hunk 4 above)
Comment (pavol47) [reply CMakeLists.txt]: resolve in commit 9344a35

File: src/SearchWidget.cpp [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+    highLightedWords(false);
+}
+
+
+void SearchWidget::keyPressEvent(QKeyEvent * event)
+{
+    if (event->key() == Qt::Key_Return || event->key() == Qt::Key_Enter) {
+        onNextButtonClicked();
+    }
+}
```
[the remark this answers is not in the document]
Comment (pavol47) [reply src/SearchWidget.cpp]: I hope that I resolve in commits f64f3e6 and 56b0c43

File: src/SearchWidget.cpp (hunk 1 above)
Comment (pavol47) [reply src/SearchWidget.cpp]: resolve in commit 17c4d5b

File: src/MainWindow.cpp (hunk 5 above)
Comment (pavol47) [reply src/MainWindow.cpp]: resolve in commit 683bfa0

File: src/SearchWidget.cpp (hunk 2 above)
Comment (pavol47) [reply src/SearchWidget.cpp]: resolve in commit baad493

**Comment by pavol47** [comment]
I compared translations and seems to me that they are same as yours. I resolved all your request but in request : " This can be simplified by just connecting QLineEdit::returnPressed() to guiNextButton::animateClick(). " I don't know if I did it right.

**Comment by agateau** [comment]
It looks good now, I am merging it. Thanks a lot for your work :+1:

**Pull request merged**